php-frontend/src/main/java/org/sonar/php/parser/LexicalConstant.java

```diff
@@ -123,10 +123,10 @@
   /**
    * Integer
    */
-  private static final String DECIMAL = "[1-9][0-9]*+|0";
-  private static final String HEXADECIMAL = "0[xX][0-9a-fA-F]++";
-  private static final String OCTAL = "0[0-7]++";
-  private static final String BINARY = "0b[01]++";
+  private static final String DECIMAL = "[1-9][0-9]*(_[0-9]+)*|0";
+  private static final String HEXADECIMAL = "0[xX][0-9a-fA-F]+(_[0-9a-fA-F]+)*";
+  private static final String OCTAL = "0[0-7]+(_[0-7]+)*";
```

Comment:
The following emoji octal is valid with my php 7.4.0RC6 but this regular expression fails to parse it: `0_0`
I would change the regular expression to: `0[0-7]*(_[0-7]+)*`
And we should always use `++` or `*+` in regular expression when possible.
So I would write: `0[0-7]*+(_[0-7]++)*+`

Reply:
Note: the following expressions are not valid with my php: `0x_0`   `0b_0`